plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,569 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE ImpredicativeTypes #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE MultiWayIf         #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE TupleSections      #-}
+{-# LANGUAGE TypeApplications   #-}
+
+module Bitwise (
+  integerToByteString,
+  byteStringToInteger,
+  andByteString,
+  iorByteString,
+  xorByteString,
+  complementByteString,
+  popCountByteString,
+  testBitByteString,
+  writeBitByteString,
+  findFirstSetByteString,
+  shiftByteString,
+  rotateByteString,
+  ) where
+
+import Control.Monad (foldM, when)
+import Control.Monad.State.Strict (State, evalState, get, modify, put)
+import Data.Bits (FiniteBits, bit, complement, popCount, rotate, shift, shiftL, xor, zeroBits, (.&.), (.|.))
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.ByteString.Unsafe (unsafePackMallocCStringLen, unsafeUseAsCString, unsafeUseAsCStringLen)
+import Data.Foldable (foldl', for_)
+import Data.Functor (void)
+import Data.Kind (Type)
+import Data.Text (Text, pack)
+import Data.Word (Word64, Word8)
+import Foreign.C.Types (CChar, CSize)
+import Foreign.Marshal.Alloc (mallocBytes)
+import Foreign.Ptr (Ptr, castPtr, plusPtr)
+import Foreign.Storable (Storable (peek, poke, sizeOf))
+import GHC.Exts (fromList, fromListN)
+import GHC.IO.Handle.Text (memcpy)
+import PlutusCore.Builtin.Emitter (Emitter, emit)
+import PlutusCore.Evaluation.Result (EvaluationResult (EvaluationFailure))
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+{-# NOINLINE rotateByteString #-}
+rotateByteString :: ByteString -> Integer -> ByteString
+rotateByteString bs i
+  -- If a ByteString is completely homogenous, rotating won't change it. This
+  -- also covers emptiness, since empty ByteStrings are homogenous vacuously.
+  | isAllZero bs || isAllOne bs = bs
```

Comment:
It's mostly that. Additionally, these checks are _much_ faster than a rotation would be, since they can work over potentially very large blocks (at least 64 bits, possibly even 64 _bytes_ if you want to pedal-to-the-metal optimize), but rotations potentially have to check every single _bit_, and then rebuild a copy of the exact same thing you already have.